Teach diagnostics to correct margin of multiline messages

estebank opened the pull request:
Make the suggestion list have a correct padding:

```
error[E0308]: mismatched types
 --> file.rs:3:20
  |
3 |     let x: usize = "";
  |                    ^^ expected usize, found reference
  |
  = note: expected type `usize`
  = note:    found type `&'static str`
  = help: here are some functions which might fulfill your needs:
          - .len()
          - .foo()
          - .bar()
```

estebank committed:
Teach diagnostics to have correctly padded lists

Make the suggestion list have a correct padding:

```
error[E0308]: mismatched types
 --> file.rs:3:20
  |
3 |     let x: usize = "";
  |                    ^^ expected usize, found reference
  |
  = note: expected type `usize`
  = note:    found type `&'static str`
  = help: here are some functions which might fulfill your needs:
          - .len()
          - .foo()
          - .bar()
```

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

KalitaAlexey commented on src/librustc_errors/emitter.rs:
```diff
@@ -923,14 +924,28 @@ impl EmitterWriter {
+                                format!("{}\n{}",
+                                        &child.message,
+                                        &child.list.iter().map(|item| {
+                                            format!("{}         - {}",
```
It is hard for to understand the inner loop.
Can you describe it?

estebank replied:
Added comments.

KalitaAlexey commented on src/librustc_errors/emitter.rs:
```diff
@@ -923,14 +924,28 @@ impl EmitterWriter {
+                                                          .collect::<String>(),
+                                                    item)
+                                        }).collect::<Vec<String>>()
+                                        .join("\n"))
```
Are you sure that `collect` and `join` are more efficient than `fold`?

KalitaAlexey commented on src/librustc_errors/emitter.rs:
```diff
@@ -699,6 +699,7 @@ impl EmitterWriter {
                     .to_string(),
                 span: MultiSpan::new(),
                 render_span: None,
+                list: vec![],
```
@nikomatsakis,
Is `vec![]` idiomatic way to create an empty vector?

KalitaAlexey commented on src/librustc_errors/diagnostic.rs:
```diff
@@ -191,11 +197,23 @@ impl Diagnostic {
            message: &str,
            span: MultiSpan,
            render_span: Option<RenderSpan>) {
+        self.sub_with_list(level, message, span, render_span, vec![]);
```
@nikomatsakis,
Is `vec![]` idiomatic way to create an empty vector?
I mean that `vec!` is a macro.
It requires expansion.
I think that `Vec::new` is more appropriate.

estebank committed:
Use fold instead of collect/join and add comments

estebank committed:
Teach diagnostics to correct margin on multiline messages

Make any diagnostic line to have the correct margin to align with the
first line:

```
error: message
 --> file.rs:3:20
  |
3 |     <CODE>
  |      ^^^^
  |
  = note: this is a multiline
          note with a correct
          margin
  = note: this is a single line note
  = help: here are some functions which might fulfill your needs:
          - .len()
          - .foo()
          - .bar()
  = suggestion: this is a multiline
                suggestion with a
                correct margin
```

estebank commented:
I realized that this could be generalized in a better way by making the diagnostic renderer aware of multiline strings, and make it provide the appropriate margin instead. By doing it that way

* the diagnostics do not need to care about lists at all
* multiline strings now should look much better (not that there're many at this time)
* the json output remains untouched, showing the unpadded multiline string

KalitaAlexey commented on src/librustc_errors/emitter.rs:
```diff
@@ -703,6 +703,21 @@ impl EmitterWriter {
+    fn msg_with_padding(&self, msg: &str, padding: usize) -> String {
+        let padding = (0..padding)
+            .map(|_| " ")
+            .collect::<String>();
```
If I get it right, it will allocate every time the method is called.
What do you think about adding a function
```rust
fn add_padding(msg: &mut String, padding: usize) {
    for _ in 0..padding {
        msg.push(' ');
    }
}
```
and using it instead of pushing padding.

KalitaAlexey commented on src/librustc_errors/emitter.rs:
```diff
@@ -703,6 +703,21 @@ impl EmitterWriter {
+            }
+            acc.push_str(&x.1);
+            acc.push_str("\n");
+            acc
```
I need other opinions, but I think that before 711 lines you call measure a new string size and reserve it to prevent allocations in 713, 715, 716 lines.

durka commented on src/librustc_errors/emitter.rs:
```diff
@@ -721,7 +736,10 @@ impl EmitterWriter {
             buffer.append(0, ": ", Style::NoStyle);
-            buffer.append(0, msg, Style::NoStyle);
+
+            // The extra 9 ` ` is the padding that's always needed to align to the `note: `.
```
How about `"note".len() + 5` (and `"suggestion".len() + 5` further down) to have one magic number instead of two?

estebank replied:
Done.

estebank committed:
Remove magic number

KalitaAlexey commented on src/librustc_errors/emitter.rs:
```diff
@@ -721,7 +736,10 @@ impl EmitterWriter {
-            buffer.append(0, msg, Style::NoStyle);
+
+            // The extra 3 ` ` is the padding that's always needed to align to the `note: `.
+            let message = self.msg_with_padding(msg, max_line_num_len + "note: ".len() + 3);
```
What does `3` stand for?

estebank replied:
```
error: message
 --> file.rs:3:20
  |
3 |     <CODE>
  |      ^^^^
  |
  = note: multiline
          message
+^^^------
|  |     |
|  |     length of "note: "
|  the magic `3`
`max_line_num_len`
```

estebank replied:
[Added this documentation to the code](https://github.com/rust-lang/rust/pull/38916/commits/04e4a60b45ec4debd20be2327cb6859271502c05).

estebank committed:
Deduplicate and document logic

nikomatsakis commented on src/librustc_errors/emitter.rs:
```diff
@@ -703,6 +703,40 @@ impl EmitterWriter {
 
+    /// Add a left margin to every line but the first, given a padding length and the label being
+    /// displayed.
+    fn msg_with_padding(&self, msg: &str, padding: usize, label: &str) -> String {
```
It seems like the labels attached to a span might also need wrapping, not just notes, right?

But I guess this code is not about "autowrap" but specifically targeting `\n` that are embedded in a message. I see, carry on.

nikomatsakis commented:
@bors r+

This is very nice.

bors commented:
:pushpin: Commit 04e4a60 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 04e4a60b45ec4debd20be2327cb6859271502c05 -->

bors commented:
:hourglass: Testing commit 04e4a60 with merge e57f061...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1498), [status-travis](https://travis-ci.org/rust-lang/rust/builds/190850945)
Approved by: nikomatsakis
Pushing e57f061be20666eb0506f6f41551c798bbb38b60 to master...
